Add support for multiple pkey algorithms

jenshenny opened the pull request:
### TODO

- [x] Read from a RSA/DSA/EC key file 
- [x] Generate a RSA/DSA/EC key 
- [x] Build and sign a gem with a RSA/DSA/EC key
- [x] Resign a gem with a RSA/DSA/EC key
- [x] Install a gem that is signed with a RSA/DSA/EC key 
        (check_trust seems to break with EC)
```
ERROR:  While executing gem ... (Gem::Security::Exception)
    trusted root certificate /CN=jenny.shen/DC=shopify/DC=com checksum does not match signing root certificate checksum
```
- [x] Sign another certificate with RSA/DSA/EC keys (no changes needed in the flow)

### Further Refinements
Create abstraction for a key (because of the API differences between RSA/DSA and EC)?

deivid-rodriguez committed:
Improve error messages in gem helpers

Previously they were printing the original command that was run, and
telling the user to rerun it. However, the command sometimes would not
match the exact command that was run (for example, when using the
`--local` flag), and in any case, it's simpler and more useful to print
the underlying error anyways.

deivid-rodriguez committed:
Reuse `sh` helper

deivid-rodriguez committed:
Simplify some code

This method always receives an array, and we require `shellwords`
unconditionally at the top of the file, so `shelljoin` will always be
available.

deivid-rodriguez committed:
Reuse `sh` helper for `git push` too

deivid-rodriguez committed:
Remove unnecessary method

deivid-rodriguez committed:
Amend changelog dates

Releases happened one day later.

deivid-rodriguez committed:
Simplify the incomplete locked specs for platform check

It doesn't really need converged specs, since it's only about the
lockfile.

deivid-rodriguez committed:
Extract `locked_dependencies` helper

deivid-rodriguez committed:
No need to use converged dependencies either

This is exclusively about the lockfile.

deivid-rodriguez committed:
Fix `bundle install` crash due to an incorrectly incomplete resolve

In case we have a corrupted lockfile that claims to support a platform, but
it's missing platform specific gems for it, bundler has a check that
detects the situation and forces a re-resolve. The result of this check
is kept under the `@locked_specs_incomplete_for_platformn` instance
variable in `Definition`.

The installer, however, calls `Definition#nothing_changed?` before this
instance variable has been filled, so the result of it is actually
incorrect here since it will claim that nothing has changed, but
something has changed (locked specs are incomplete for the current
platform).

The consequence of this incorrect result is that the installer thinks it
can go on without re-resolving, resulting in the incomplete resolution
from the lockfile being used, and in a crash being triggered due to
that.

The solution is to make sure the `@locked_specs_incomplete_for_platform`
instance variable is filled before `nothing_changed?` gets called.
Moving it to `initialize` makes the most sense, not because it's the
best place for it (we can refactor this later), but because all of the
other "outdated definition" checks are already set there.

deivid-rodriguez committed:
Refactor symlink attack specs

deivid-rodriguez committed:
Don't make duplicated checks on directory safety

This seems to speed up `gem install` on Windows by more than 50%.

deivid-rodriguez committed:
Simplify `File.expand_path` usage

deivid-rodriguez committed:
Remove redundant `File.expand_path`

`File.realpath` already expands paths.

deivid-rodriguez committed:
Install location safety should consider casing

deivid-rodriguez committed:
Rename a couple of workflows for consistency

deivid-rodriguez committed:
Fix typo in job name

deivid-rodriguez committed:
Run CI against JRuby 9.3.0.0

deivid-rodriguez committed:
Check safety of packaged symlinks

If we explicitly disallow the creation of symlinks that point to files
outside of the destination directory, we can avoid any other safety
checks while creating directories, because we can be sure they will
always fall under the destination directory as well.

deivid-rodriguez committed:
We no longer need to check realpaths

Since symlinks and absolute paths are already checked.

deivid-rodriguez committed:
Explicit check file not created outside of destination

jenshenny committed:
Add support to build and sign certificates with multiple key algorithms

deivid-rodriguez committed:
Changelog for Bundler version 2.2.29

deivid-rodriguez committed:
Changelog for Rubygems version 3.2.29

deivid-rodriguez committed:
Fix typo

We want to update the original repo, not a fresh one.

This went undetected because the `git commit` command inside the
`update_git` method ignores failures, and in this case it was failing
because all files are untracked in the new repo.

I will fix that later but for now fix the typo.

deivid-rodriguez committed:
Keep different code in custom branch in dummy repo

This doesn't affect the outcome of the test, but it makes the `git
commit` command inside `update_git` not fail because of not having
anything to commit.

deivid-rodriguez committed:
Make adding files and committing into a single command

For efficiency.

deivid-rodriguez committed:
Let `update_git` work with whatever branch is checkout out

Unless the `:branch` option is passed.

It's more efficient, and it results in less hardcoding of "master".

deivid-rodriguez committed:
`git commit` no longer needs to be allowed to fail

deivid-rodriguez committed:
I see no harm in this, allow it

deivid-rodriguez committed:
Merge pull request #4969 from rubygems/cherry_pick_changelogs

Changelogs for rubygems 3.2.29 and bundler 2.2.29

deivid-rodriguez committed:
Merge pull request #4970 from rubygems/simplify_dummy_git_repo_handling

Simplify dummy git repo handling in bundler specs

doodzik commented on lib/rubygems/commands/cert_command.rb:
```diff
@@ -170,7 +175,8 @@ def build_key # :nodoc:
 
-    key      = Gem::Security.create_key
+    algorithm = options[:key_algorithm] || 'RSA'
+    key      = Gem::Security.create_key(options[:key_algorithm])
```
```suggestion
    key      = Gem::Security.create_key(algorithm)
```

doodzik commented on lib/rubygems/security/signer.rb:
```diff
@@ -141,7 +139,7 @@ def sign(data)
     raise Gem::Security::Exception, 'no certs provided' if @cert_chain.empty?
 
-    if @cert_chain.length == 1 and @cert_chain.last.not_after < Time.now
+    if true # @cert_chain.length == 1 and @cert_chain.last.not_after < Time.now
```
Why did this change?

jenshenny replied:
I wanted to get into the resign code path. Alternatively, I could have created an expired cert (which I should've done anyways). Feel free to reset the last commit I pushed btw, it was mostly for debugging.

There are two areas that need to be changed in the resign path
1. the `re_sign_key` method in the signer class. It is called when the cert is expired when signing a gem.[`OpenSSL::PKey::RSA.new(File.read(disk_key_path), @passphrase) rescue nil`](https://github.com/jenshenny/rubygems/blob/master/lib/rubygems/security/signer.rb#L181 )
2. [`Gem:Security.re_sign`](https://github.com/jenshenny/rubygems/blob/master/lib/rubygems/security.rb#L495) which resigns the expired cert (called by `re_sign_key` above)

doodzik commented on lib/rubygems/security/policy.rb:
```diff
@@ -163,6 +165,7 @@ def check_trust(chain, digester, trust_dir)
       raise Gem::Security::Exception, message
     end
 
+    # TODO: check if this still works
```
```suggestion
```

deivid-rodriguez committed:
Show the exact name of the gem that was deleted

If a non exact name (matched as a regexp) is passed to `bundle info`,
these strings might not match.

deivid-rodriguez committed:
Show a warning in `bundle info` if gem has been deleted

deivid-rodriguez committed:
Match a more exact warning message

deivid-rodriguez committed:
Get specs green on arm64-darwin-20

deivid-rodriguez committed:
Make spec more realistic

The spec was just faking an installed gemspec without any installed gem
backing it up, resulting in `bundle install` claiming that the gem was
already installed when it was not.

deivid-rodriguez committed:
Fix `bundle install` to force reinstallation of deleted gems

nobu committed:
Needs time.rb for `Time#iso8601`

`Gem::Request.verify_certificate_message` used this method without
"time" which defines it.

ruby/ruby@b16d876ac2d20b3c17a6906860d3a3f279aecf65

deivid-rodriguez committed:
Merge pull request #4975 from nobu/time-in-request

Gem::Request.verify_certificate_message needs time.rb for iso8601

deivid-rodriguez committed:
Merge pull request #4960 from rubygems/speed_up_gem_install

Speed up `gem install`, specially under Windows

deivid-rodriguez committed:
Prefer `require_relative` for internal requires

deivid-rodriguez committed:
Avoid loading the `digest` gem unnecessarily

OpenSSL includes what we need.

deivid-rodriguez committed:
Merge pull request #4977 from rubygems/get_specs_green_on_arm64_darwin_20

Get specs green on arm64-darwin-20

simi committed:
Merge pull request #4978 from rubygems/more_relative_requires

Prefer `require_relative` for all internal requires

deivid-rodriguez committed:
Merge pull request #4971 from rubygems/bundle_info_on_deleted_gem

Make `bundle info` show a proper warning every time it finds a deleted gem

deivid-rodriguez committed:
Remove comment that seems no longer accurate

We require things inline or autoload them for efficiency and to avoid as
much as possible conflicting with user's choice of gems.

But I removed the require and observed no deadlocks when activating
gems.

deivid-rodriguez committed:
Unify issue template and ISSUES.md document

Some crucial information to ease maintainers work, like the advice of
upgrading rubygems and bundler, was one step away from the issue
template, making it easier for some users to miss.

Now all relevant information is written directly in the bug report
template.

deivid-rodriguez committed:
Remove unnecessary code

All supported rubygems versions implement this.

deivid-rodriguez committed:
Use correct way to detect default gems

The other way, in particular matching a substring in the gemspec
summary, is brittle and no longer used since Ruby 2.0.

This needed rewriting the specs that depended on that way.

deivid-rodriguez committed:
Merge pull request #4980 from rubygems/further_improve_issue_template

Unify issue template and ISSUES.md document

deivid-rodriguez committed:
Merge pull request #4974 from rubygems/bundle_install_reinstalls_deleted_gems

Fix `bundle install` to reinstall deleted gems

deivid-rodriguez committed:
Merge pull request #4979 from rubygems/dont_load_digest

Avoid loading the `digest` gem unnecessarily

deivid-rodriguez committed:
Merge pull request #4941 from rubygems/fix_install_crash

Fix install crash when lockfile has missing dependencies for the current platform

deivid-rodriguez committed:
Merge pull request #4962 from rubygems/jruby_9300

Run CI against JRuby 9.3.0.0

nobu committed:
Keep loaded features

Now `$LOADED_FEATURES` list is being maintained by `setup` and
`teardown` and, only libaries under the temporary directory will
be removed.  As `save_loaded_features` removes the rest libraries
other than this test directory, ordinary libraries loaded from
files under rubygems also removed, and often causes constant
redefinition warnings.

deivid-rodriguez committed:
Merge pull request #4872 from rubygems/better_gem_helper_errors

Print better errors when `bundler/gem_tasks` fail

nobu committed:
Remove save_loaded_features

deivid-rodriguez committed:
Merge pull request #4983 from nobu/keep-loaded_features

Keep loaded features

deivid-rodriguez committed:
Show a proper error if gem path is not writable

Instead of showing the bug report place with an error at a randome
place.

deivid-rodriguez committed:
Show proper error when previous installation of gem can't be deleted

Instead of showing the bug report template with an error at a random
place.

deivid-rodriguez committed:
Catch up with recent error message modification

deivid-rodriguez committed:
Test only first and last rubies supported upstream on MacOS

deivid-rodriguez committed:
Merge pull request #4965 from rubygems/permission_stuff

Give better errors for some permission issues

yakschuss committed:
Update broken link in Bundler::Fetcher::CertificateFailureError

simi committed:
Merge pull request #4987 from yakschuss/update-broken-link-in-fetcher-error

Update broken link in Bundler::Fetcher::CertificateFailureError

jenshenny commented on lib/rubygems/security/policy.rb:
```diff
@@ -164,9 +166,9 @@ def check_trust(chain, digester, trust_dir)
     end
 
     save_cert = OpenSSL::X509::Certificate.new File.read path
-    save_dgst = digester.digest save_cert.public_key.to_s
+    save_dgst = digester.digest save_cert.public_key.to_pem
 
-    pkey_str = root.public_key.to_s
+    pkey_str = root.public_key.to_pem
```
Context: [EC keys](https://ruby-doc.org/stdlib-2.5.1/libdoc/openssl/rdoc/OpenSSL/PKey/EC.html#method-i-to_text) don't have `to_s` method like [RSA](https://ruby-doc.org/stdlib-2.5.1/libdoc/openssl/rdoc/OpenSSL/PKey/RSA.html).

rochlefebvre commented on lib/rubygems/security.rb:
```diff
@@ -337,17 +337,14 @@ class Exception < Gem::Exception; end
 
   ##
-  # Length of keys created by KEY_ALGORITHM
+  # Key used for Elliptic Curve
```
I think this is acually a "named curve", or something along those lines.

rochlefebvre commented on lib/rubygems/security.rb:
```diff
@@ -337,17 +337,14 @@ class Exception < Gem::Exception; end
+  # Key used for Elliptic Curve
 
-  KEY_LENGTH = 3072
+  EC_KEY = 'secp384r1'
```
Any reasons for choosing this particular curve as the default? Also, thoughts on exposing a parameter to specify a different curve?

doodzik replied:
From what I can tell the difference is [negligible for the use case here](https://security.stackexchange.com/questions/78621/which-elliptic-curve-should-i-use).

rochlefebvre commented on lib/rubygems/security.rb:
```diff
@@ -461,9 +469,23 @@ def self.create_digest(algorithm = DIGEST_NAME)
 
-  def self.create_key(length = KEY_LENGTH, algorithm = KEY_ALGORITHM)
-    algorithm.new length
+  def self.create_key(algorithm)
```
This is a public method, and we should expect that there's code out there that calls `Gem::Security.create_key` in its current form, using any permutation of the existing parameters.

Perhaps we should define a new method and call that one in the `cert` command instead?

doodzik replied:
I don't think this is an issue because we changed all occurrences in the codebase and I assume it isn't likely for someone else to depend on it. Let's see what the people from RubyGems say.

jenshenny commented on lib/rubygems/security/signer.rb:
```diff
@@ -83,8 +83,8 @@ def initialize(key, cert_chain, passphrase = nil, options = {})
-    if @key && !@key.is_a?(OpenSSL::PKey::RSA)
-      @key = OpenSSL::PKey::RSA.new(File.read(@key), @passphrase)
+    if @key && !@key.is_a?(OpenSSL::PKey::PKey)
+      @key = OpenSSL::PKey.read(File.read(@key), @passphrase) if @key
```
```suggestion
      @key = OpenSSL::PKey.read(File.read(@key), @passphrase)
```

doodzik commented:
Opened the PR on rubygems.
`https://github.com/rubygems/rubygems/pull/4991`
